### PR TITLE
[MTP][Sparse MLA] Take advantage of native MTP support in indexer when possible

### PR DESCRIPTION
**Pull request by MatthewBonanni**
## Purpose
PR #34552 added support for MTP > 1 with sparse MLA by unconditionally flattening all requests into single-token decodes. The indexer kernel does support MTP = 1, though, and future iterations of the kernel will support other token counts e.g. MTP = 3. This PR takes advantage of this support by only flattening when the specified `num_speculative_tokens` is not natively supported by the kernel.

It does this by turning on `require_uniform` when taking this pathway. When very short prefills (< `1 + num_speculative_tokens`) are present in the batch, these requests and all others below (including decodes) will be treated as prefills. This is suboptimal, but expected to be very rare. Addressing it would require reordering the batch into [speculative decodes, short prefills, all other prefills] rather than just [decodes, prefills], which would add complexity.

## Test Plan
```
vllm serve deepseek-ai/DeepSeek-V3.2 \
    -tp 8 -ep \
    --speculative-config '{"method": "mtp", "num_speculative_tokens": 1}' \
    --no-enable-prefix-caching
```
with
```
vllm bench serve \
    --dataset-name spec_bench \
    --dataset-path question.jsonl \
    --spec-bench-output-len 1024 \
    --seed 42 \
    --ignore-eos \
    --temperature 0 \
    --tokenizer deepseek-ai/DeepSeek-V3
```

### Main
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Benchmark duration (s):                  96.72     
Total input tokens:                      310052    
Total generated tokens:                  1024000   
Request throughput (req/s):              10.34     
Output token throughput (tok/s):         10587.74  
Peak output token throughput (tok/s):    9616.00   
Peak concurrent requests:                1000.00   
Total token throughput (tok/s):          13793.54  
---------------Time to First Token----------------
Mean TTFT (ms):                          7737.62   
Median TTFT (ms):                        7051.60   
P99 TTFT (ms):                           15349.34  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          72.98     
Median TPOT (ms):                        72.92     
P99 TPOT (ms):                           81.29     
---------------Inter-token Latency----------------
Mean ITL (ms):                           139.33    
Median ITL (ms):                         113.82    
P99 ITL (ms):                            368.19    
---------------Speculative Decoding---------------
Acceptance rate (%):                     91.04     
Acceptance length:                       1.91      
Drafts:                                  535608    
Draft tokens:                            535608    
Accepted tokens:                         487602    
Per-position acceptance (%):
  Position 0:                            91.04     
==================================================
```
### PR
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Benchmark duration (s):                  95.90     
Total input tokens:                      310052    
Total generated tokens:                  1024000   
Request throughput (req/s):              10.43     
Output token throughput (tok/s):         10678.25  
Peak output token throughput (tok/s):    9786.00   
Peak concurrent requests:                1000.00   
Total token throughput (tok/s):          13911.46  
---------------Time to First Token----------------
Mean TTFT (ms):                          7779.73   
Median TTFT (ms):                        7314.22   
P99 TTFT (ms):                           15318.20  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          72.71     
Median TPOT (ms):                        72.83     
P99 TPOT (ms):                           80.71     
---------------Inter-token Latency----------------
Mean ITL (ms):                           138.74    
Median ITL (ms):                         113.15    
P99 ITL (ms):                            366.24    
---------------Speculative Decoding---------------
Acceptance rate (%):                     90.96     
Acceptance length:                       1.91      
Drafts:                                  535831    
Draft tokens:                            535831    
Accepted tokens:                         487380    
Per-position acceptance (%):
  Position 0:                            90.96     
==================================================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: csrc/sampler.cu [hunk 1]
```diff
@@ -575,7 +575,7 @@ static __global__ __launch_bounds__(kNumThreadsPerBlock) void topKPerRowDecode(
   // The range of logits within the row.
   int rowStart = 0;
   int seq_len = seqLens[rowIdx / next_n];
-  int rowEnd = seq_len - next_n + (rowIdx % next_n) + 1;
+  int rowEnd = max(0, seq_len - next_n + (rowIdx % next_n) + 1);
```
Comment (benchislett) [inline csrc/sampler.cu:578]: What's this for?

Comment (MatthewBonanni) [reply csrc/sampler.cu:578]: Cudagraph padding requests with `seq_len == 0`. We could alternatively clamp `seq_lens` to a minimum of `next_n` on the python side but the kernel will do dummy work

**Review by LucasWilkinson — APPROVED** [review]
LGTM; thanks for following up on this!
